### DISCUSSION TITLE
Is this the expected behavior when running in worker mode?

### DISCUSSION THREAD

**Discussion by twisted1919** [php/frankenphp#1250] [Q&A] [answered]
I have an interesting problem where a certain header, which is set only in a search area, once set, it will be part of all requests.
I am using API-Platform and Symfony, latest versions for both of them, in production.
Also using `1.3-php8.3-alpine` for frankenphp.

In a state provider, I am doing something like:
```php
public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null
    {
        if ($operation instanceof CollectionOperationInterface) {
            ...

            $this->responseCustomHeaders->add([
                'X-JobSearch-Id' => (int) Carbon::now()->timestamp,
            ]);

            ...

            return new TraversablePaginator(
               ...
            );
        }

        return [];
    }
```
Once set, the `X-JobSearch-Id` is returned in all requests. 
`responseCustomHeaders` is just a regular service. 

Is the state of this service shared between all requests? Should I reset these custom headers when I am done with them? 
This looks like a bug TBH, but wanted to ask first.

Thanks.

**Comment by dunglas** [comment]
This looks intended. Your service must implement the Symfony `ResetInterface` it it has states that must be reset between requests. What is even better is to design services as stateless.

Would you mind to share the code of this service?

**Reply by twisted1919** [reply]
Here's the service:
```php
<?php

declare(strict_types=1);

namespace App\Service;

use Symfony\Component\HttpFoundation\HeaderBag;

final readonly class ResponseCustomHeaderBag implements ResponseCustomHeaderBagInterface
{
    public function __construct(
        private HeaderBag $headers = new HeaderBag(),
    ) {
    }

    /**
     * @param array<string,mixed> $headers
     */
    #[\Override]
    public function add(array $headers): void
    {
        $this->headers->add($headers);
    }

    /**
     * @return array<string,mixed>
     */
    #[\Override]
    public function all(): array
    {
        return $this->headers->all();
    }
}

```
> Your service must implement the Symfony ResetInterface it it has states that must be reset between requests.   

I did not know about this, is there any specific docs for this in relation to frankenphp?

> What is even better is to design services as stateless.  

That's not always possible.

**Reply by dunglas** [reply, marked as answer]
Indeed the service is the problem. I don't why you need a service here instead of attaching the custom' headers directly to the current request.

The `ResetInterface` is not FrankenPHP-specific (it is also useful for Messenger, Swoole etc) but indeed we should add some docs for it in this context (here is the existing docs: https://symfony.com/doc/current/messenger.html#stateless-worker).
While convenient, it's almost always possible to design stateless services instead (which is a best practice anyway).

**Reply by twisted1919** [reply]
@dunglas - yeah, I don't know why that service was like that, I have removed them and used the request stack. Moreover, I used [phanalist](https://github.com/denzyldick/phanalist) to analize and expose services with state, which shown about 2 of them and I corrected them as well by implementing the `ResetInterface` on them.
This seems to have fixed the problem, and I agree with you, the docs could do a bit better here, a warning at least.

Anyway, thank you!
